### PR TITLE
Alm 1045 dockerize

### PR DESCRIPTION
**Pull request by JackLaBarba**
https://jira.plos.org/jira/browse/ALM-1045

## Related PRs:

https://github.com/PLOS/lagotto/pull/40
https://github.com/PLOS/molten/pull/1256

**Note** that this PR merges into a branch called `aws` instead of dev. This PR breaks compatibility with trusty--I felt like it was too complicated and low value to maintain this in Soma. The `aws` branch is meant to be the basis of what we'll create `PLOS/lagotto-fomula` from (heads up @heschlie).

## What's here?

* Run puma inside a container
* Run nginx inside a container
* Reworked how map.jinja is translated into ENV vars for the container
* Got rid of a lot of stuff that was rendered unnecessary by containers
* Set up consul service discovery
* Moved db seeding concerns from these states into the lagotto repo. Seeding no longer happens automatically (it seems like it's best done manually)

## What Consul services are advertised now?
( https://confluence.plos.org/confluence/display/SRE/Service+Naming+Convention )

* alm-manager-memcache
* alm-manager-sql
* alm-manager-nosql (kinda, this couch is only managed by these salt states in monolithic environments)
* alm-manager-redis
* alm-manager-app
* alm-manager-web

## How to test this

* check out this molten PR: https://github.com/PLOS/molten/pull/1256
* change the image tag name at https://github.com/PLOS-Formulas/lagotto-formula/pull/19/files#diff-bc9ee8db26e65f08826be6c6cd1009dfR49 from `develop` to `ALM-1045-dockerize` (Once https://github.com/PLOS/lagotto/pull/40 merges, this change will be unnecessary)
* `vagrant up` from `molten/vagrant/apps/lagotto`
* You'll see twenty-something errors (from what I can tell, these are race conditions arising from docker and consul setup)
* `vagrant provision` (should run clean)
* check it out at http://lagotto.vagrant.local

### CODE REVIEW

**Comment by JackLaBarba** [comment]
closed in favor of https://github.com/PLOS/lagotto-formula/pull/1

**Comment by JackLaBarba** [comment]
Nevermind, getting this to work with pogos2 is a whole separate can of worms that we probably shouldn't open until this here worm can is closed.

File: lagotto/init.sls [hunk 1]
```diff
@@ -1,123 +1,76 @@
...
   - lagotto.common
-{% if grains['fqdn'] == sidekiq_server %}
-  - lagotto.sidekiq
-{% endif %}
+  - lagotto.lib.docker-common
 
-{% set app_root = props.get('app_root') %}
-{% set ruby_ver = props.get('version_ruby') %}
+{% set environment = salt.grains.get('environment') %}
+{% set app_name = 'lagotto' %}
```
Comment (sgatchell) [inline lagotto/init.sls]: some of these assignments look like they're duplicated in `docker-common.sls`

Comment (JackLaBarba) [reply lagotto/init.sls]: you right, I DRYed up this stuff in 6be8437

File: README.md [hunk 2]
```diff
@@ -1 +1,7 @@
 Salt formula for Lagotto
+
+## A note about Docker images in Vagrant
+
+By default, these states will use the lagotto image tagged `develop` in vagrant.
+If you would like to experiement with another image, change the
+`docker_image_tag` value under `vagrant` in `map.jinja`.
```
Comment (sgatchell) [inline README.md:7]: Good to note! I hope we implicitly standardize on this approach.

File: lagotto/init.sls [hunk 3]
```diff
@@ -1,123 +1,69 @@
...
+{{ app_name }}-network:
+  docker_network.present:
+    - name: {{ app_name }}
+
+{{ app_name }}-app-container-running:
+  docker_container.running:
+    - name: {{ app_name }}-app
+    - image: {{ docker_image_name }}
+    - environment:
+{%- for name, value in props.items() %}
```
Comment (sgatchell) [inline lagotto/init.sls:35]: Simple translation!  I like the idea of expecting salt props to be exposed as env vars.

**Review by sgatchell — APPROVED** [review]
It took me an extra provision, after the recommended provision, but the app is up now in my vagrant.  

This is a great cleansing transformation! :shipit:

File: lagotto/init.sls (hunk 3 above)
Comment (JackLaBarba) [reply lagotto/init.sls:35]: Yeah! I wasn't quite sure how controversial this idea would be. It sure makes these states cleaner and easier to work with, but it comes with the downside that _everything_ in this dict will be made available to the docker container, even stuff it doesn't strictly need to know about.

**Comment by JackLaBarba** [comment]
> It took me an extra provision, after the recommended provision, but the app is up now in my vagrant.

Right, was there a single errorred state about the removing the railsassets volume? I still haven't figured that one out. It seems more like a problem between salt and docker than a problem between me and salt frankly. I did what I could to make sure Salt's not trying to remove a volume that doesn't exist, but it still has that little initializing hiccup.

I don't love that the error's in there, but there's probably more important things for me to be sorting out right now.

Thanks for the :shipit: ! I also need a review over on https://github.com/PLOS/molten/pull/1256 if you don't mind and https://github.com/PLOS/lagotto/pull/40 needs a final thumbs up (again) before it can mere.

File: lagotto/init.sls (hunk 3 above)
Comment (sgatchell) [reply lagotto/init.sls:35]: A whitelist would be simple too, though it doesn't seem controversial enough with lagotto to warrant it.

**Pull request merged**